src/layout.rs

```diff
@@ -0,0 +1,295 @@
+        self._step
+    }
+}
+
+// Rescale so that pos in [-scale, scale].
+fn rescale(pos: &mut Vec<Point>, scale: Nt, indices: Vec<usize>) {
+    let n = indices.len();
+    if n == 0 {
+        return;
+    }
+    // find mean in each dimension
+    let mut mu: Point = [0.0, 0.0];
+    for &n in &indices {
+        mu[0] += pos[n][0];
+        mu[1] += pos[n][1];
+    }
+    mu[0] /= n as Nt;
+    mu[1] /= n as Nt;
+
+    // substract mean and find max coordinate for all axes
+    let mut lim = std::f64::NEG_INFINITY;
+    for n in indices {
+        let [px, py] = pos.get_mut(n).unwrap();
+        *px -= mu[0];
+        *py -= mu[1];
+
+        let pm = max(px.abs(), py.abs());
+        if lim < pm {
+            lim = pm;
+        }
+    }
+
+    // rescale
+    for [px, py] in pos.iter_mut() {
+        *px *= scale / lim;
+        *py *= scale / lim;
+    }
+}
+
+fn recenter(pos: &mut Vec<Point>, center: Point) {
+    for [px, py] in pos.iter_mut() {
+        *px += center[0];
+        *py += center[1];
+    }
+}
+
+#[allow(clippy::too_many_arguments)]
+pub fn evolve<Ty, Fa, Fr, C>(
+    graph: &Graph<Ty>,
+    mut pos: Vec<Point>,
+    fixed: HashSet<usize>,
+    f_a: Fa,
+    f_r: Fr,
+    mut cs: C,
+    niter: usize,
+    tol: f64,
+    weights: HashMap<(usize, usize), f64>,
+    scale: Option<Nt>,
+    center: Option<Point>,
+) -> Vec<Point>
+where
+    Ty: EdgeType,
+    Fa: Force,
+    Fr: Force,
+    C: CoolingScheme,
+{
+    let mut step = cs.update_step(std::f64::INFINITY);
+
+    for _ in 0..niter {
+        let mut energy = 0.0;
+        let mut converged = true;
+
+        for v in graph.node_indices() {
+            let v = v.index();
+            if fixed.contains(&v) {
+                continue;
+            }
+            // attractive forces
+            let ys = graph.neighbors_undirected(NodeIndex::new(v)).map(|n| {
+                let n = n.index();
+                (&pos[n], weights[&(v, n)])
+            });
+            let fa = f_a.total(&pos[v], ys);
+
+            // repulsive forces
+            let ys =
+                graph.node_indices().filter(|&n| n.index() != v).map(|n| {
+                    let n = n.index();
+                    (&pos[n], 1.0)
+                });
+            let fr = f_r.total(&pos[v], ys);
+
+            // update current position
+            let f = [fa[0] + fr[0], fa[1] + fr[1]];
+            let f2 = f[0] * f[0] + f[1] * f[1];
+            energy += f2;
+
+            let fnorm = max(f2.sqrt(), LBOUND);
+            let dx = step * f[0] / fnorm;
+            let dy = step * f[1] / fnorm;
+            pos[v][0] += dx;
+            pos[v][1] += dy;
+
+            if dx * dx + dy * dy > tol {
+                converged = false;
+            }
+        }
+
+        step = cs.update_step(energy);
+        if converged {
+            break;
+        }
+    }
+
+    if fixed.is_empty() {
+        if let Some(scale) = scale {
+            rescale(
+                &mut pos,
```

Comment:
To match the behavior of networkx we need to set scale == 1.0 instead of not calling it when `scale` is `None`, because rescale always moves the mean of points to (0, 0).

Reply:
i match this behavior by setting `scale=1.0` by default in `spring_layout` function in `retworkx/__init__.py` . i was thinking that it makes sense to explicitly define `scale=None` if you don't want any rescaling of the output.

Reply:
ah yeah, thanks for clarifying! I thought it might be confusing for the user that `scale = 1.0` and `scale = None` have different behaviors, but yeah it's not that big a problem and yours makes complete sense too. I think maybe it's just better to not use `Option` and set `scale = 1.0` in `retworkx/__init__.py`, so passing `None` wouldn't be allowed